src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs

```diff
@@ -13,17 +13,15 @@ internal abstract class OSFileStreamStrategy : FileStreamStrategy
     {
         protected readonly SafeFileHandle _fileHandle; // only ever null if ctor throws
         private readonly FileAccess _access; // What file was opened for.
-        private readonly FileShare _share;
 
         protected long _filePosition;
         private long _appendStart; // When appending, prevent overwriting file.
-        private long _length = -1; // When the file is locked for writes on Windows (_share <= FileShare.Read) cache file length in-memory, negative means that hasn't been fetched.
-        private bool _exposedHandle; // created from handle, or SafeFileHandle was used and the handle got exposed
+        private long _length = -1; // When the file is locked for writes on Windows ((share & FileShare.Write) == 0) cache file length in-memory, negative means that hasn't been fetched.
+        private bool _exposedHandle; // created from handle, or SafeFileHandle was used and the handle got exposed, or FileShare.Write was specified when the handle was opened.
```

Comment:
It seems that we should change the name of this field. Historically it meant "the handle has been exposed, validate file offset on every async operaiton":

https://github.com/dotnet/runtime/blob/50413eafd5a930dc5f3cce1a7fcba743cbb36c80/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.cs#L301-L303

Now it means "don't cache file length as it's not safe".

```suggestion
        private bool _lengthCanBeCached;
```